### PR TITLE
libgcrypt: fix bad locking behavior introduced in update

### PR DESCRIPTION
**Pull request by dtzWill**
As-is this hangs on musl (and possibly cause of Darwin hangs)
due to trying to obtain non-recursive lock already held.

Especially if this seems to fix hanging problems on Darwin,
will be reporting upstream.

Probably should wait until that happens or this is
otherwise investigated further.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dtzWill** [comment]
Looks like yes hangs w/o the patch--time to reconfirm with patch once timeout triggers!

**Comment by dtzWill** [comment]
I suppose this should go to staging first due to mass-rebuilds...

**Comment by dtzWill** [comment]
...But that's messy because staging doesn't have the commit from master/18.03 that disables tests that this fixes.  Blargh.

**Comment by LnL7** [comment]
@dtzWill let's merge master into staging first then.

**Comment by dtzWill** [comment]
> @dtzWill let's merge master into staging first then.

Yes please.  Having trouble getting any response from upstream:
* tried sending to mailing list, pending moderator approval
* tried subscribing to mailing list (so messages go through), similarly blocked on approval
* tried opening account on their bug tracker and a)offered twitter sign-in is broken and b)also requires approval which hasn't happened yet.

Might just submit as a pull request which isn't the preferred method but I don't want to spend days waiting on this, I'd like to move on :).

**Comment by dtzWill** [comment]
Hooray they approved my phrabricator account, lol.

**Comment by dtzWill** [comment]
Accepted upstream, committed to master and 1.8 branch! \o/

**Comment by LnL7** [comment]
Great, let's send this to staging then. The conflict should be resolved now.

**Comment by dtzWill** [comment]
Targeting staging now, think this is good to go?

**Review by LnL7 — APPROVED** [review]
LGTM if ofborg agrees.

**Comment by LnL7** [comment]
@GrahamcOfBorg build libgcrypt

**Pull request merged**
